Added package x264

pretyman opened the pull request:
Hi Ruslan,

Added package x264 based on Autotools which generates pkg-config modules. 
I wrote a routine to create CMake targets from pkg-config modules through find scripts to make life easier to use them with CMake.

Tell me what you think.

ruslo commented on cmake/modules/hunter_finalize.cmake:
```diff
@@ -86,6 +86,10 @@ macro(hunter_finalize)
+
+    # Automatically generated Find sripts from pkg-config modules
+    # go into ${HUNTER_INSTALL_PREFIX}/lib/cmake, so:
+    list(APPEND CMAKE_MODULE_PATH "${HUNTER_INSTALL_PREFIX}/lib/cmake")
```
Let's use `${HUNTER_INSTALL_PREFIX}/cmake/Hunter` here.

Just for your info:
* http://cgold.readthedocs.io/en/latest/tutorials/install/layout.html

pretyman replied:
Ok, cool, thanks

ruslo commented on cmake/projects/x264/hunter.cmake:
```diff
@@ -0,0 +1,34 @@
+hunter_cacheable(x264)
+hunter_cmake_args(
+    x264
+    CMAKE_ARGS         # do not use double quotes on CMAKE_ARGS
```
> do not use double quotes on CMAKE_ARGS

Why not?

This should work for example:
```cmake
hunter_cmake_args(
    x264
    CMAKE_ARGS
      "EXTRA_FLAGS=--disable-asm"
      "PKGCONFIG_EXPORT_TARGETS=x264"
)
```

pretyman replied:
But this won't:

"PKGCONFIG_EXPORT_TARGETS=x264;other-pkg-config-module2;other-pkg-config-module3"

nor

PKGCONFIG_EXPORT_TARGETS="x264;other-pkg-config-module2;other-pkg-config-module3"

We tried before ;-)

ruslo replied:
> "PKGCONFIG_EXPORT_TARGETS=x264;other-pkg-config-module2;other-pkg-config-module3"

Should work

> PKGCONFIG_EXPORT_TARGETS="x264;other-pkg-config-module2;other-pkg-config-module3"

Yes, because it's unquoted argument

> We tried before ;-)

Yep, I remember. This is the reason I have this section:
* http://cgold.readthedocs.io/en/latest/tutorials/variables/regular.html#quotes

pretyman replied:
> Should work

Thanks, it does.

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,113 @@
+    string(COMPARE NOTEQUAL
+        "${@PKG_CONFIG_MODULE@_INCLUDE_DIRS}"
+        ""
+        _@PKG_CONFIG_MODULE@_has_include_dirs
```
I think `_is_empty` will work just fine here.

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,113 @@
+          ${@PKG_CONFIG_MODULE@_LDFLAGS}
+      )
+    endif()
+    message("AWP: ${_@PKG_CONFIG_MODULE@_link_libs}")
```
`hunter_status_debug` or `hunter_print_status`

pretyman replied:
ooops, sorry, that slipped :S was supposed to remove that

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,113 @@
+    # --- INTERFACE_COMPILE_OPTIONS end ---
+
+  elseif(@PKG_CONFIG_MODULE@_FIND_REQUIRED)
+    message(FATAL_ERROR "Could not find pkg-config module: @PKG_CONFIG_MODULE@")
```
`hunter_internal_error` I guess

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,113 @@
+    string(COMPARE NOTEQUAL
+        "${@PKG_CONFIG_MODULE@_CFLAGS_OTHER}"
+        ""
+        _@PKG_CONFIG_MODULE@_has_cflags_other
```
To avoid such long ugly names you can wrap everything with function.

pretyman replied:
Sorry, I guess I was overly verbose, I did not understand what you meant, but I am just doing to change for _is_empty as you suggested above

ruslo replied:
> I did not understand what you meant

Declare the function in this module and use it. Since function has it's own scope we can use nice short names:

```cmake
function(hunter_pkgconfig_export_targets ...)
  string(COMPARE EQUAL "..." "" has_ldflags)
  set(link_libs "") # we can use pretty name here
  if(has_ldflags)
    list(APPEND link_libs ${${PKG_CONFIG_MODULE}_LDFLAGS_OTHER})
  endif()
endfunction()

hunter_pkgconfig_export_targets("@PKG_CONFIG_MODULE@")
```

pretyman replied:
Ok, so we would split this file into two:

1. cmake/modules/hunter_pkgconfig_export_targets.cmake
2. scripts/pkgconfig-export-targets.cmake.in

Is this understanding correct?

ruslo replied:
> Is this understanding correct?

Yes

ruslo replied:
> Declare the function in this module and use it

I guess it's better to declare the function in separate module as you suggested.

pretyman replied:
> Declare the function in this module and use it

I actually skipped that part when reading for some reason Ô.o

pretyman commented:
Updated this pull request.

I ended up opting to not create the file `cmake/modules/hunter_pkgconfig_export_targets.cmake` and leaving the function in the file `pkgconfig-export-targets.cmake.in` as you first suggested, because the `modules` directory it getting a little overcrowded (it has qt related modules, so I guess it would be a good idea in the future to put project related modules in the project directories).

Please review

ruslo commented:
> Updated this pull request

Thanks, reviewing.

> because the modules directory it getting a little overcrowded

I see no problem about it.

> it has qt related modules, so I guess it would be a good idea in the future to put project related modules in the project directories

Path to modules added to `CMAKE_MODULE_PATH`. Even if they will live in different directories, logically they will be still in the same place. If all modules live in one directory we have "natural" protection from collusion - you can't have two different modules with the same name because you can't have two files with the same name in directory.

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,133 @@
+  endif()
+endfunction()
+
+if(NOT TARGET "@PKG_CONFIG_MODULE@")
```
Condition can be moved to the function.

pretyman replied:
Done

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,133 @@
+  set(MODULE_REQUIRED ${ARGV1})
+
+  pkg_check_modules(${PKG_CONFIG_MODULE} ${PKG_CONFIG_MODULE})
+  if(${PKG_CONFIG_MODULE}_FOUND)
```
```cmake
if(NOT ..._FOUND)
  hunter_internal_error("...")
endif()
```
will save us indented space.

ruslo commented on scripts/pkgconfig-export-targets.cmake.in:
```diff
@@ -0,0 +1,133 @@
+if(NOT TARGET "@PKG_CONFIG_MODULE@")
+  hunter_pkgconfig_export_target(
+    "@PKG_CONFIG_MODULE@"
+    @PKG_CONFIG_MODULE@_FIND_REQUIRED
```
Since module installed by Hunter I guess `REQUIRED` is always `TRUE`?

pretyman replied:
That's true, yeah, I can remove that

pretyman commented:
Ok, updated :+1:

ruslo commented on cmake/modules/hunter_pkgconfig_export_target.cmake:
```diff
@@ -0,0 +1,126 @@
+
+function(hunter_pkgconfig_export_target)
+  set(PKG_CONFIG_MODULE ${ARGV0})
+  if(NOT TARGET "${PKG_CONFIG_MODULE}")
```
```cmake
if(TARGET "${PKG_CONFIG_MODULE}")
  return()
endif()

# -2 indent here
```

pretyman replied:
:+1:

pretyman commented:
Updated, now I think now we are good to go

ruslo commented:
Testing:
* https://travis-ci.org/ingenue/hunter/builds/237027853?utm_source=github_status&utm_medium=notification

ruslo commented:
Package building with GCC even if Clang set in toolchain:
* https://travis-ci.org/ingenue/hunter/jobs/237027855

From log:
```
Unknown option CC=/usr/local/clang-3.5.0/bin/clang, ignored
Unknown option CXX=/usr/local/clang-3.5.0/bin/clang++, ignored
```

pretyman commented:
Thanks for pointing it out. 

Actually, I forgot I have some local changes to `modules/hunter_autotools_project` where I do the configure command line with the toolchain information as variables before the `./configure` command and not after. 

To be clear, I do:
`CC=.... CXX=... ./configure`  (1)

instead of:
`./configure CC=... CXX=...` (2)

This package `x264` is not a real autotools package, although autotools should work with (1), I will have to confirm in the docs if (1) can be kept instead of (2) that we have today.

I hope the explanation was clear.

pretyman commented:
Hum, not very clear https://www.gnu.org/software/automake/manual/html_node/Standard-Configuration-Variables.html

ruslo commented:
`./configure` of x264 package is not generated by autotools, it's a manually written script as far as I understand.

pretyman commented:
True, but Autotools should read from variables as well put before the `./configure` script as well in theory. I see that the `AR` flag was put before the `./configure` ( https://github.com/ruslo/hunter/blob/c8f92f5e2bf3a024af7b416ff43f6e538638d0f5/cmake/modules/hunter_autotools_project.cmake#L312 ) but I can't find the reason (the line was reformatted on https://github.com/ruslo/hunter/commit/e3baf868061718359bf3922fcd6a8df8d5e67d7c#diff-7e0239d435b39e7d38f47d30ecb93bf6L248 - and I don't know how to look at that line before this commit )

I mean, we could put the toolchain information before and after the `./configure` to be sure, it would just be noisy on the debug message

pretyman commented:
Well, the Autotools packages for odb and SQLite3 all work with the flags before the `./configure`

pretyman commented:
Updated this pull request with the change in `hunter_autotools_project`

pretyman commented:
meh, actually this is no good still, the same should be done do CPPFLAGS, CFLAGS, CXXFLAGS (+others)

pretyman commented:
hum, ok I need some help here :>

I think my options are:
1. create a new scheme for this package, like this `hunter_autotools_project` wouldn't be touched (but if I don't touch it, there could be a lot of repetition and we won't get DRY
2. refactor `hunter_autotools_project` to put these flags as variables before the command line which should be supported by Autotools anyway. The thing is that I don't have access to OSX nor an iPhone to test the iOS build :/

Do you have any suggestions Ruslan?

Sorry, this should have been a simple package, but it turned out to be a little monster on its own

ruslo commented:
> The thing is that I don't have access to OSX nor an iPhone to test the iOS build :/

You can use Travis, we have OSX/iOS toolchains in testing.

> Do you have any suggestions Ruslan?

Nope don't have any. The (2) approach looks like the only available. Just for your info issue with using `CC=... ./some-script` in `ExternalProject` stages: https://github.com/ruslo/hunter/issues/246

> Sorry, this should have been a simple package, but it turned out to be a little monster on its own

Happens quite often for non-CMake packages :)

pretyman commented:
Hi Ruslan,

Updated this pull request:
I ended up not changing the behaviour of the current autotools build (I guess it was safer this way)
So I refactored out into functions behaviour that would be useful when integrating other packages with custom build systems such as x264 (and FFmpeg if I end up neededing it).

What I did was:

1. created function `hunter_get_toolchain_binaries` to return the the toolchain binaries
2. created function `hunter_get_build_flags` which gets the complete flags for a given configuration type, adding the -I<root-it>/include and -L<root-it>/lib to the respective flags
2. created function `hunter_autotools_configure_command` which sets up the configure command for the package, reducing the size of the `hunter_autotools_project`  and separating the creating of this command for when there is a solution for running the commands in a specific environment, right now I'm still using `VAR=VAL ./configure...` tecnique because I tried to implement a way respecting https://github.com/ruslo/hunter/issues/246 but it became non trivial and it was best to discuss it first

pretyman commented:
After the review I will comment the function headers and update this pull request, I just didn't want to write stuff that could end up being undone.

Thank

pretyman commented:
I have some concerns about the vocabulary of the parameters and function names, what we talk about `<root-id>` here in comments is called `GLOBAL_INSTALL_DIR` in the parameters, would be nice to have a common term for this to avoid confusion

ruslo commented on cmake/configs/default.cmake:
```diff
@@ -209,6 +209,7 @@ hunter_config(tinydir VERSION 1.2-p0)
 hunter_config(websocketpp VERSION 0.7.0-p2)
 hunter_config(wxWidgets VERSION 3.0.2)
 hunter_config(x11 VERSION 1.5.0)
+hunter_config(x264 VERSION snapshot-20170420-2245)
```
We need an example to test this package.

ruslo commented on cmake/projects/x264/schemes/url_sha1_x264.cmake.in:
```diff
@@ -0,0 +1,341 @@
+  set(HUNTER_PACKAGE_INSTALL_PREFIX "@HUNTER_PACKAGE_INSTALL_PREFIX@")
+  configure_file(
+      "@HUNTER_GLOBAL_SCRIPT_DIR@/${merge_lipo_script}.in"
+      "@HUNTER_PACKAGE_SOURCE_DIR@/universal/${merge_lipo_script}"
```
I'm afraid file from this directory will be removed on ExternalProject_Add stage. That's why we need to configure to `HUNTER_PACKAGE_BUILD_DIR` and then copy from `HUNTER_PACKAGE_BUILD_DIR` on EP stage (see Qt).

pretyman replied:
I have not tested the iOS build, but I find it odd that this works https://github.com/ruslo/hunter/blob/79197dd68de9fdeb2d5761a7e4d994baf431232f/cmake/modules/hunter_autotools_project.cmake#L501-L506 (this is where I took this code from), care to explain how does that one works? (I agree it shouldn't, maybe it is a bug on CMake for not clearing the directory before executing the EP stages?)

ruslo replied:
> care to explain how does that one works?

Let's test it. With what package does it work?

pretyman replied:
Ah, ok, I think I understand why it works. `@HUNTER_PACKAGE_SOURCE_DIR@/universal` is used as a source directory of a `ExternalProject_Add(...)` call without a file to be extracted, and it is before extracting the file that the directory is removed. I guess in case one wants to use a path as a source directory, which seems to be legal according to the documentation: https://cmake.org/cmake/help/v3.0/module/ExternalProject.html .

`If SOURCE_DIR is explicitly set to an existing directory the project will be built from it(...)` 

I skimmed through the source of `ExternalProject.cmake` and it seems that it is before extracting the file of an EP that the directory is removed.

In the iOS build, it happens that we use `@HUNTER_PACKAGE_SOURCE_DIR@/multi-arch-build/${ios_architecture}` for each iOS architecture as the source of the EP calls with the package files to be extracted, so the `@HUNTER_PACKAGE_SOURCE_DIR@/universal` is never touched. So I think we are safe.

Since you had this doubt in while reviewing it shows that this code is rather non-trivial, I actually tried to extract this behavior into a function to be reused in other packages when integrating other build systems, but found it rather hard. First need to treat the running commands with specific env vars, then I guess this can be looked upon.

> Let's test it. With what package does it work?

`odb-sqlite` and its dependencies `SQLite3` and `odb` used to work

ruslo replied:
> its dependencies SQLite3 and odb used to work

I don't see iOS toolchain in testing:
* https://travis-ci.org/ingenue/hunter/builds/156129910
* https://travis-ci.org/ingenue/hunter/builds/146957970

pretyman replied:
Yeah, true, it was commented out because we had no support for *.pc files at the time, should work now, I'll uncomment the ios toolchain and send a pr

pretyman replied:
Aye Caramba, actually these packages are not working on iOS upstream, I had code to locally treat `*.pc` files, which I don't think I sent upstream :S 

We had a problem with unrelocatable files during the merge, `*.pc` text files had in them the directory of the install prefix, so the `autotools-merge-lipo` script [ https://github.com/ruslo/hunter/blob/bfdc19731a5e467674abd7aef9a227678c4176a6/scripts/autotools-merge-lipo.cmake.in ] would fail merging them.

By looking at the script now, I don't think I uploaded the logic to hunter repo.

Sorry about the confusion, this was done some time ago. I had a prototype of an app using `odb-sqlite` on iOS hence the confusion, I don't think these packages are working with the upstream version of hunter.

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -98,7 +98,7 @@ function(hunter_autotools_project target_name)
   if(PARAM_UNPARSED_ARGUMENTS)
     hunter_internal_error(
-        "Invalid arguments passed to hunter_autotools_configure:"
+        "Invalid arguments passed to hunter_autotools_project:"
```
We need to retest at least one project that was using this function, like `odb-boost`. In this case INTERNAL_ID of `odb-boost` should be changed.

pretyman replied:
What do you think of https://github.com/ruslo/hunter/issues/837 ? I can bump `odb-boost`, but me feeling is that it wastes disk space on the cache server since the packages shouldn't change. Anyway, I'll do as you see best fit

ruslo replied:
This is the only requirement for now, if you update INTERNAL_ID I will run tests and merge this pull request.

pretyman replied:
Ok, thanks, updated the pull request

pretyman commented:
Updated this PR with the example, sorry I did this in a hurry and forgot ;-)

pretyman commented:
Updated this PR by bumping odb-boost's PACKAGE_INTERNAL_DEPS_ID

ruslo commented on cmake/modules/hunter_pkgconfig_export_target.cmake:
```diff
@@ -0,0 +1,128 @@
+  if(TARGET "${PKG_CONFIG_MODULE}")
+    return()
+  endif()
+  pkg_check_modules(${PKG_CONFIG_MODULE} ${PKG_CONFIG_MODULE})
```
This function will find packages from Hunter because it respects CMAKE_PREFIX_PATH variable. But this functionality available since CMake 3.1:
* https://cmake.org/cmake/help/v3.1/module/FindPkgConfig.html
```
the CMAKE_PREFIX_PATH, CMAKE_FRAMEWORK_PATH, and
CMAKE_APPBUNDLE_PATH cache and environment variables will be
added to pkg-config search path
```
Not available in CMake 3.0:
* https://cmake.org/cmake/help/v3.0/module/FindPkgConfig.html

By the way further documentation reading, since CMake 3.6 we have imported targets from the box:
* https://cmake.org/cmake/help/v3.6/module/FindPkgConfig.html

```
The IMPORTED_TARGET argument will create an imported
target named PkgConfig::<PREFIX>> that can be passed directly
as an argument to target_link_libraries()
```

We have to use `cmake_minimum_required(VERSION)` or `if(CMAKE_VERSION VERSION_LESS)` here.

pretyman replied:
> This function will find packages from Hunter because it respects CMAKE_PREFIX_PATH variable.

We changed `PKG_CONFIG_LIBDIR` so it would find packages inside the Hunter build tree: 
https://github.com/ruslo/hunter/blob/bfdc19731a5e467674abd7aef9a227678c4176a6/cmake/modules/hunter_finalize.cmake#L85-L88 so I believe this should work, but did it fail in the tests?

> By the way further documentation reading, since CMake 3.6 we have imported targets from the box:

Doh, so it seems like I wasted my time on this? :P 

Maybe use `if(CMAKE_VERSION VERSION_LESS)` as you proposed and export the target under `PkgConfig::` to have the same behavior?

pretyman replied:
Using cmake 3.8.0-rc2 with `IMPORTED_TARGET`, it did not link properly:
```
CMakeFiles/x264-example.dir/example.c.o: In function `main':
example.c:(.text+0xc8): undefined reference to `x264_param_default_preset'
example.c:(.text+0x11b): undefined reference to `x264_param_apply_profile'
example.c:(.text+0x144): undefined reference to `x264_picture_alloc'
example.c:(.text+0x15b): undefined reference to `x264_encoder_open_148'
example.c:(.text+0x281): undefined reference to `x264_encoder_encode'
example.c:(.text+0x312): undefined reference to `x264_encoder_encode'
example.c:(.text+0x369): undefined reference to `x264_encoder_delayed_frames'
example.c:(.text+0x380): undefined reference to `x264_encoder_close'
example.c:(.text+0x38f): undefined reference to `x264_picture_clean'
example.c:(.text+0x3af): undefined reference to `x264_encoder_close'
example.c:(.text+0x3c1): undefined reference to `x264_picture_clean'
collect2: error: ld returned 1 exit status
```

While using hunter version it worked fine.

ruslo replied:
Can you compare the link lines? Probably it's a valid CMake issue.

pretyman replied:
Yea, link lines are above, but for clarity, `IMPORTED_TARGET`:

```
/usr/bin/gcc   -rdynamic CMakeFiles/x264-example.dir/example.c.o  -o x264-example
```

Hunter's version:
```
/usr/bin/gcc   -rdynamic CMakeFiles/x264-example.dir/example.c.o  -o x264-example -L/home/discipline/Development/CMake/hunter/_Base/xxxxxxx/bd0d80f/ddaf47f/Install/lib -lx264 -lpthread -lm -ldl
```

ruslo replied:
So all dependencies are missing... Let's forget about this for now, we can always improve this later.

pretyman replied:
I just debugged CMake's version of the `IMPORTED_TARGET` and yes, it is bugged.

It takes the flags and tries to convert them into full paths to the files with `find_library`.

pkg-config LDFLAGS start with: `-L<root-id>/lib -lx264 -lpthread -lm -ldl`.
CMake's IMPORTED target takes the first argument `<root-id>/lib` and uses as a directory to call `find_library(...)` on the libraries `x264`, `pthread`, `m`, `dl`. The result will be added to the imported target's libraries. However it finds only `x264`, but not others which are present in the toolchain, they become `pthread-NOTFOUND`, `m-NOTFOUND`, `dl-NOTFOUND`. it it append these values to a var `_libs`, and below checks `if(_libs) add libs to target`, where this if returns false because of the `NOTFOUND` elements (last element?). So it does not add any libs to the target.

Possible fix: if `find_library(...)` returns `lib-NOTFOUND` then use the given pkg-config's linking flags `-lpthread`, etc.. instead of trying to link to the absolute path of the lib.

Hope this explanation was clear :-)

ruslo replied:
Okay, so we can use your version for now. Also you can report CMake issue :)

pretyman replied:
Cool, thanks

ruslo commented on cmake/schemes/url_sha1_autotools.cmake.in:
```diff
@@ -58,3 +58,11 @@ hunter_autotools_project("@HUNTER_EP_NAME@"
+foreach(PKG_CONFIG_MODULE ${PKGCONFIG_EXPORT_TARGETS})
+  configure_file(
+      "@HUNTER_GLOBAL_SCRIPT_DIR@/pkgconfig-export-targets.cmake.in"
+      "@HUNTER_PACKAGE_INSTALL_PREFIX@/cmake/Hunter/Find${PKG_CONFIG_MODULE}.cmake"
```
Well, I just figured out that if you will install `${PKG_CONFIG_MODULE}Config.cmake` to `lib/cmake/${PKG_CONFIG_MODULE}/${PKG_CONFIG_MODULE}Config.cmake` then we don't need to modify `hunter_finalize` and we can use `find_package(<PKG> CONFIG REQUIRED)`. Sorry for the late thought :)

pretyman replied:
Ok, I can put them into `${PKG_CONFIG_MODULE}Config.cmake` so it is less code into `hunter_finalize`

pretyman commented:
Hi Ruslan,

Updated this pull request:
1. Removed the creation of the `lib/cmake/${PKG_CONFIG_MODULE}/${PKG_CONFIG_MODULE}Config.cmake`, in favor of just calling `hunter_pkgconfig_export_target(target)` to create a `PkgConfig::target` target, then we can use the CMake's version when it gets fixed inside the function. Creating such file was just bloat, really. IMO calling the function is better API than the previous approach of `find_package(...)` because to understand find_package's result one always need to look up the docs on what is going to happen :-)
2. Basic comments on the new functions

ruslo commented:
> Creating such file was just bloat, really. IMO calling the function is better API than the previous approach of find_package(...) because to understand find_package's result

I don't think it's a good idea. `hunter_add_package(foo)` + `find_package(foo CONFIG REQUIRED)` is a clean and simple CMake-friendly pattern. 

> one always need to look up the docs on what is going to happen

What document? Why it's better than looking to the Hunter docs for what is `hunter_pkgconfig_export_target` instead of standard CMake `find_package`?

pretyman commented:
> What document?

The docs for the `find_package` script one wants to use. Example: after `find_package(Boost)` it writes to variable `Boost_LIBRARIES`. Yes I understand you advocate the `CONFIG` of Boost when using with hunter, but I'm just trying to show the general use case of `find_package`. So then there is `find_package(x11)`, can we expect it to write the libraries to `x11_LIBRARIES`? Nope, it writes to `X11_LIBRARIES`. So `find_package(OpenSSL)` will write the libs to `OpenSSL_LIBRARIES`? Not too fast, it writes to `OPENSSL_LIBRARIES`

And `find_package(Threads)` does something different.

So now `find_package(x264)` is going to do something else.

I feel that `find_package` is not consistent, making it time consuming to work with it.

> Why it's better than looking to the Hunter docs for what is hunter_pkgconfig_export_target instead of standard CMake find_package

Because you would look just once.  `hunter_pkgconfig_export_target` exports a target under `PkgConfig::pkg-config-module` or errors out if it doesn't find the module.

> hunter_add_package(foo) + find_package(foo CONFIG REQUIRED) is a clean and simple CMake-friendly pattern.

I agree that it is a CMake-friendly pattern, it is just that I feel that CMake is not friendly :P

Obivously I would do as you say. :-)

ruslo commented:
> I feel that find_package is not consistent, making it time consuming to work with it

I still don't understand how custom `hunter_pkgconfig_export_target` is better. Now user have to answer "does this package use find_package or hunter_pkgconfig_export_target?". You have to look at the `pkg.*` wiki page anyway.

> Because you would look just once

With `find_package(x264 CONFIG REQUIRED)` you have to look once too. Why do you need second time? :)

> I agree that it is a CMake-friendly pattern, it is just that I feel that CMake is not friendly :P

Well, this should be improved in CMake first, then Hunter will adopt it. Hunter is CMake-based and for CMake, so it's reasonable to use best CMake practices.

> Obivously I would do as you say

Let's use `find_package`. I think it's important to use unified approach everywhere. Just for your info `find_package` can work with package which is a bunch of CMake modules (see [pkg.gauze](https://github.com/ruslo/hunter/wiki/pkg.gauze)). It would be good to make an example of all approaches from [this page](https://docs.hunter.sh/en/latest/overview/anything.html). Another idea (just an idea) with hooks in `find_package` described here: https://github.com/ruslo/hunter/issues/812#issuecomment-305465004

pretyman commented:
> You have to look at the pkg.* wiki page anyway.

That is true for when you are writing from scratch.

> I still don't understand how custom hunter_pkgconfig_export_target is better. 

Because you see it once, you can predict side effects of this call, for example: if you read `hunter_pkgconfig_export_target(swscale)`, you know that it will create `PkgConfig::swscale`. But when you read `find_package(swscale)` you can't tell the side effects of this call.

But it is ok, I will revert the changes to create the `<pkg-config-module>Config.cmake` as before.

ruslo commented:
>  But when you read find_package(swscale) you can't tell the side effects of this call

Because one project can have several targets, `find_package(swscale)` can introduce `swscale::one` and `swscale::two` imported targets.

> if you read hunter_pkgconfig_export_target(swscale), you know that it will create PkgConfig::swscale

I think I've missed this part, probably it should be `swscale::swscale`.

pretyman commented:
> Because one project can have several targets

But for pkg-config it is one target per pkg-config module.

> I think I've missed this part, probably it should be swscale::swscale

I was trying to mimic CMake's `IMPORTED_TARGET`, so we can have a version check inside `hunter_pkgconfig_export_target` and use CMake's own function, which exports to `PkgConfig::swscale`.

ruslo commented:
> I was trying to mimic CMake's IMPORTED_TARGET

Yes, okay

pretyman commented:
Pull request updated!

Let's move forward! \o/

ruslo commented on cmake/modules/hunter_get_build_flags.cmake:
```diff
@@ -0,0 +1,154 @@
+      )
+    endforeach()
+
+    # -> CMAKE_C_FLAGS
```
I think this comment doesn't belong here.

pretyman replied:
You are right ;-)

ruslo commented on examples/x264/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+project(x264-example)
+
+hunter_add_package(x264)
+find_package(x264)
```
`find_package(x264 CONFIG REQUIRED)`

pretyman commented:
Updated the pull request with your requested.

May the tests run /o/

ruslo commented:
odb-boost:
* https://travis-ci.org/ingenue/hunter/builds/247564933?utm_source=github_status&utm_medium=notification

x264:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1884
* https://travis-ci.org/ingenue/hunter/builds/247565494?utm_source=github_status&utm_medium=notification

ruslo commented:
> x264

* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1886
* https://travis-ci.org/ingenue/hunter/builds/247568274?utm_source=github_status&utm_medium=notification

ruslo commented:
> odb-boost

```
CMake Error at /home/travis/build/ingenue/hunter/_testing/Hunter/_Base/Download/Hunter/unknown/a3f7e09/Unpacked/cmake/modules/hunter_autotools_project.cmake:60 (include):

  include could not find load file:

    hunter_autotools_set_toolchain_binaries
```

pretyman commented:
Ooops, sorry about that, updated the pull request :-)

Let the good tests roll \o\

ruslo commented:
> x264

Some tests failed, feel free to exclude.

pretyman commented:
> Some tests failed, feel free to exclude.

Ok, but something is odd with the Android toolchain though, it seems to be missing the C preprocessor binary:
https://travis-ci.org/ingenue/hunter/jobs/247568280

ruslo commented:
> Ok, but something is odd with the Android toolchain though, it seems to be missing the C preprocessor binary

Hunter moved to CMake 3.7 based Android toolchain and variables. CMAKE_C_PREPROCESSOR is not defined by CMake. Discussion:
* https://cmake.org/pipermail/cmake-developers/2016-November/029598.html

pretyman commented:
O.o

I wish they had some converter from autotools to CMake to import existing projects. Developgeting is so much fun, but getting stuck at these buildings issues it is rather annoying.

Ok, so we have to set `CMAKE_C_PREPROCESSOR` to `${CMAKE_C_COMPILER} -E` somewhere. I have to download the Android NDK and test it locally.

(on a side note, I think `hunter_get_build_flags` has similar responsibilities from `hunter_dump_cmake_flags`, should  they be merged into `hunter_dump_flags`?

ruslo commented:
> on a side note, I think hunter_get_build_flags has similar responsibilities from hunter_dump_cmake_flags, should they be merged into hunter_dump_flags

The purpose of the `hunter_dump_cmake_flags` is to convert cmake properties like `CMAKE_POSITION_INDEPENDENT_CODE` into real compiler flags and save these flags to `CMAKE_CXX_FLAGS` for futher usage in build scheme. I guess the `get_directory_property` part can be moved there too.

pretyman commented:
How are the other autotools packages for Android working for 3.7 CMake? Since I see it Android is enabled for `pkg.odb` I guess either tests didn't run on updating the toolchain or there must have been a change somewhere to allow it to run and I missed it on the refactoring of `hunter_autotools_project` inside this present pull request

ruslo commented:
Looks good:
* https://travis-ci.org/ingenue/hunter/builds/248054179?utm_source=github_status&utm_medium=notification

ruslo commented:
https://github.com/ruslo/hunter/blob/1b23acfca02d7a8aba8404431eeb1e5901e89c60/cmake/modules/hunter_autotools_project.cmake#L119-L132

pretyman commented:
Ah, thanks for the heads up, updated the pull request

ruslo commented:
x264:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1894
* https://travis-ci.org/ingenue/hunter/builds/248449649?utm_source=github_status&utm_medium=notification

odb-boost:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1893
* https://travis-ci.org/ingenue/hunter/builds/248449502?utm_source=github_status&utm_medium=notification

ruslo commented:
> x264

Some toolchains failed, please take a look.

pretyman commented:
Thanks, will be commenting them

pretyman commented:
Done: https://github.com/ingenue/hunter/pull/85

Thanks for taking care of AppVeyor :-)

ruslo commented:
x264:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1900
* https://travis-ci.org/ingenue/hunter/builds/249040856?utm_source=github_status&utm_medium=notification

odb-boost:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1901
* https://travis-ci.org/ingenue/hunter/builds/249040906?utm_source=github_status&utm_medium=notification

ruslo commented:
Uploading.

x264:
* https://travis-ci.org/ingenue/hunter/builds/249046022?utm_source=github_status&utm_medium=notification
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1903

odb-boost:
* https://travis-ci.org/ingenue/hunter/builds/249045834?utm_source=github_status&utm_medium=notification
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1902

ruslo commented:
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.22

pretyman commented:
Great, thanks!

ruslo commented on cmake/projects/x264/schemes/url_sha1_x264.cmake.in:
```diff
@@ -0,0 +1,342 @@
+set(default_path "/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin")
+set(shell_env_path "PATH=@HUNTER_PACKAGE_INSTALL_PREFIX@/bin:${default_path}")
+
+set(d1 "${root_id}/lib/pkgconfig")
```
`root_id` is not defined. This code is not covered by `x264` or `odb-boost`, I can reproduce error by building `xau`.

Fixed: https://github.com/ruslo/hunter/commit/dc1a977628e84eae7469eb1d2037c61cdf23a2da
